File: leetcode/78(中等)/1. 子集.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def subsets(nums)
  (0..(2 ** nums.size - 1))
      .map { |x| x.to_s(2).rjust(nums.size, '0') }
      .map do |x|
          x.split('')
          .enum_for(:each_with_index)
          .to_a
          .map { |char, index| char.eql?('0') ? -1 : index }
          .map do |index|
              index.eql?(-1) ? '' : nums[index]
          end
          .select { |x| x != '' }
      end
      .to_a
end

File: leetcode/78(中等)/1. 子集.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    return new Array(Math.pow(2, nums.length))
      .fill(0)
      .map((_, i) => {
        const binary = fixed(toBinary(i), nums.length, '0');
        return binary
          .split('')
          .map((char, index) => char === '1' ? nums[index] : '')
          .filter(e => e !== '');
      });
};

const toBinary = (num) => {
  let binary = '';
  while (num > 0) {
    binary = (num % 2) + binary;
    num = Math.floor(num / 2);
  }
  return binary;
}

const fixed = (str, len, placeHolder) => {
  if (str.length >= len) return str;
  return placeHolder.repeat(len - str.length) + str;
}